src/snowflake/connector/config_parser/parser.py

```diff
@@ -0,0 +1,298 @@
+        _options: List of ConfigOptions that are our children.
+        _root_parser: Reference to root parser. Used to efficiently propagate to
+          child options.
+        _nest_path: The names of the ConfigParsers that this parser is nested
+          under. Used to efficiently propagate to child options.
+    """
+
+    def __init__(
+        self,
+        *,
+        name: str,
+        file_path: Path | None = None,
+    ):
+        """Create a new ConfigParser.
+
+        Args:
+            name: Name of this ConfigParser.
+            file_path: File this parser should read values from. Can be omitted
+              for all child parsers.
+        """
+        self.name = name
+        self.file_path = file_path
+        # Objects holding subparsers and options
+        self._options: dict[str, ConfigOption] = dict()
+        self._sub_parsers: dict[str, ConfigParser] = dict()
+        # Dictionary to cache read in config file
+        self.conf_file_cache: TOMLDocument | None = None
+        # Information necessary to be able to nest elements
+        #  and add options in O(1)
+        self._root_parser: ConfigParser = self
+        self._nest_path = [name]
+
+    def read_config(
+        self,
+    ) -> None:
+        """Read and cache config file."""
+        if self.file_path is None:
+            raise ConfigParserError(
+                "ConfigParser is trying to read config file, but it doesn't have one"
+            )
+        LOGGER.debug(f"reading configuration file from {str(self.file_path)}")
+        try:
+            self.conf_file_cache = tomlkit.parse(self.file_path.read_text())
+        except Exception as e:
+            raise ConfigSourceError(
+                f"An unknown error happened while loading '{str(self.file_path)}"
+                f"', please see the error: {e}"
+            )
+
+    def add_option(
+        self,
+        *,
+        option_cls: type[ConfigOption] = ConfigOption,
+        **kwargs,
+    ) -> None:
+        """Add an ConfigOption to this ConfigParser.
+
+        Args:
+            option_cls: The class that should be instantiated. This is class
+              should be a child class of ConfigOption. Mainly useful for cases
+              where the default ConfigOption needs to be extended, for example
+              if a new configuration option source needs to be supported.
+        """
+        kwargs["_root_parser"] = self._root_parser
+        kwargs["_nest_path"] = self._nest_path
+        new_option = option_cls(
+            **kwargs,
+        )
+        self._check_child_conflict(new_option.name)
+        self._options[new_option.name] = new_option
+
+    def _check_child_conflict(self, name: str) -> None:
+        """Check if a sub-parser, or ConfigOption conflicts with given name.
+
+        Args:
+            name: Name to check against children.
+        """
+        if name in (self._options.keys() | self._sub_parsers.keys()):
```

Comment:
Double check, but I don't think you need `.keys()` here as iteration over keys is usually the default.